crates/spark-wallet/src/wallet.rs

```diff
@@ -1299,11 +1334,126 @@ impl SparkWallet {
                 private_enabled: Some(private_enabled),
-                master_identity_public_key: None,
             })
             .await?;
         Ok(())
     }
+
+    /// Starts leaf optimization in the background.
+    pub fn start_leaf_optimization(&self) {
+        self.leaf_optimizer.start();
+    }
+
+    /// Cancels the ongoing leaf optimization.
+    ///
+    /// This sets a cancellation flag that is checked between rounds.
+    /// The current round will complete before stopping.
+    pub async fn cancel_leaf_optimization(&self) -> Result<(), SparkWalletError> {
+        self.leaf_optimizer.cancel().await?;
+        Ok(())
+    }
+
+    /// Returns the current optimization progress snapshot.
+    pub fn get_leaf_optimization_progress(&self) -> OptimizationProgress {
+        self.leaf_optimizer.progress()
+    }
+
+    /// Selects leaves with automatic retry logic.
+    ///
+    /// This method implements a multi-level retry strategy:
+    /// 1. Outer retry loop (up to SELECT_LEAVES_MAX_RETRIES times) handles general failures like
+    ///    attempting to pay before initial sync by refreshing leaves between retries.
+    /// 2. Inner retry logic handles insufficient funds when optimization is in progress by
+    ///    cancelling the optimization and retrying once.
+    async fn select_leaves_with_retry(
+        &self,
+        target_amounts: Option<&TargetAmounts>,
+    ) -> Result<spark::tree::LeavesReservation, SparkWalletError> {
+        use spark::tree::TreeServiceError;
+
+        let mut reservation: Option<spark::tree::LeavesReservation> = None;
+
+        for i in 0..SELECT_LEAVES_MAX_RETRIES {
+            // Try to select leaves with inner optimization retry logic
+            let reserve_result = self
+                .try_select_leaves_with_optimization_retry(target_amounts)
+                .await;
+
+            match reserve_result {
+                Ok(r) => {
+                    reservation = Some(r);
+                    break;
+                }
+                Err(e) => {
+                    error!("Failed to select leaves: {e:?}");
+                }
+            }
+
+            info!("Failed to select leaves, refreshing leaves and retrying");
+            self.tree_service.refresh_leaves().await?;
+
+            // Check if we have enough funds after refresh
+            if let Some(target_amounts) = target_amounts {
+                let available_balance = self.tree_service.get_available_balance().await?;
+                if available_balance < target_amounts.total_sats() {
+                    info!("Not enough funds to select leaves after refresh");
+                    return Err(TreeServiceError::InsufficientFunds.into());
+                }
+            }
+
+            // Sleep between retries (except on the last one)
+            if i < SELECT_LEAVES_MAX_RETRIES - 1 {
+                tokio::time::sleep(Duration::from_secs(1)).await;
+            }
```

Comment:
Should we try selecting leaves again after the refresh immediately? That might save a second.

Reply:
I actually didn't mean getting rid of the sleep, but trying another round of selecting after the refresh and before the sleep. That could go in a different PR. Or not.